# TF-IDF

We'll rely on term frequency times inverted document frequency to measure meaningful similarity between documents. Let's start by generating a matrix for the separate constituent parts of _Stjórn_.

In [1]:
import os,glob,json
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [2]:
def normalize(target):
    # This dict further standardizes the text beyond the rule set of stjorn-extract.ipynb:
    matrix = {
        'j': 'i',
        'v': 'u',
        'ð': 'þ',
        'á': 'a',
        'ǽ': 'æ',
        'é': 'e',
        'í': 'i',
        'ó': 'o',
        'ú': 'u',
        'ý': 'y',
        'ǿ': 'ø',
        'k': 'c',
        }
    for k,v in matrix.items():
        target = target.replace(k, v)
    return target

titles = ['prologue', 'introduction', 'gn', 'ex', 'lv', 'nm', 'dt', 'ios', 'idc', 'rt', '1sm', '2sm', '3rg', '4rg']
tokens = []
for title in titles:
    with open(f"nlp/{title}.txt") as raw:
        document = raw.read().replace('\n', ' ')
        tokens.extend(document.split())

work_indices = {
    'stjorn1': (650,124417),
    'stjorn2': (124417,147678),
    'stjorn3': (147678,156943,160719),
    'stjorn4': (156943,160719)
}

stjorn = dict()
for _work, _range in work_indices.items():
    if len(_range) == 2:
        stjorn[_work] = normalize(' '.join(tokens[_range[0]:_range[1]]))
    else:
        stjorn[_work] = normalize(' '.join(tokens[_range[0]:_range[1]] + tokens[_range[2]:]))

menota = dict()
for text in glob.glob('../menota/dipl/*txt'):
    ref = os.path.basename(text).replace('.txt', '')
    with open(text) as doc:
        # NB for present purposes I'm subjecting Menota to the same exaggerated normalization standard as Stjórn:
        menota[ref] = normalize(doc.read().replace('\n', ''))

In [3]:
vectorizer = TfidfVectorizer(min_df=1)
model = vectorizer.fit_transform(stjorn.values())
df = pd.DataFrame(cosine_similarity(model), stjorn.keys(), stjorn.keys())
df

stjorn1   stjorn2   stjorn3   stjorn4
stjorn1  1.000000  0.811736  0.903759  0.842739
stjorn2  0.811736  1.000000  0.797618  0.833766
stjorn3  0.903759  0.797618  1.000000  0.835339
stjorn4  0.842739  0.833766  0.835339  1.000000

After eliminating vowel length and the þ/ð distinction, these are now all pretty similar to one another, with the biggest difference between _Stjórn II_ and _III_.

Now let's first add _Konungs skuggsjá_ from Menota, as well as Unger's own edition of the _Norwegian Homily Book_. Fingers crossed that we have got the normalization standard of the former to approach Unger's methods reasonably well.

In [4]:
nhb_titles = ['alcuin', 'hom', 'olafr', 'visio', 'paternoster', 'anhang1'] # this is the sequence matched in Menota
nhb = ''
for title in nhb_titles:
    filepath = f'../nhb/nlp/{title}.txt'
    with open(filepath) as doc:
        nhb = nhb + doc.read().replace('\n', '')
stjorn_plus = []
for v in stjorn.values():
    stjorn_plus.append(v)
stjorn_plus.extend([menota['nks235g_konungs_skuggsja'], nhb])
model = vectorizer.fit_transform(stjorn_plus)
df = pd.DataFrame(cosine_similarity(model), list(stjorn.keys()) + ['ks', 'nhb'], list(stjorn.keys()) + ['ks', 'nhb'])
df

stjorn1   stjorn2   stjorn3   stjorn4        ks       nhb
stjorn1  1.000000  0.794571  0.883003  0.812112  0.705110  0.711152
stjorn2  0.794571  1.000000  0.773820  0.803064  0.690873  0.671303
stjorn3  0.883003  0.773820  1.000000  0.797769  0.739994  0.740641
stjorn4  0.812112  0.803064  0.797769  1.000000  0.656851  0.677968
ks       0.705110  0.690873  0.739994  0.656851  1.000000  0.664858
nhb      0.711152  0.671303  0.740641  0.677968  0.664858  1.000000

_Stjórn III_ and _Konungs skuggsjá_ share material cognate within the vernacular, so this connection standing out as the strongest between _Konungs skuggsjá_ and the constituent parts of _Stjórn_ comes as no surprise; if anything, the difference in match with the other parts is rather small. In fact, the _Norwegian Homily Book_ has a higher match with _Stjórn III_ than _Konungs skuggsjá_ does.

Next, let's model all of Menota along with Stjórn. Perhaps we'll leave Unger's _Homily Book_ in alongside the Menota edition, just for comparison's sake.

In [5]:
corpus = []
titles = []
for k,v in stjorn.items():
    titles.append(k)
    corpus.append(v)
titles.append('nhb')
corpus.append(nhb)
for k,v in menota.items():
    titles.append(k)
    corpus.append(v)
model = vectorizer.fit_transform(corpus)
df = pd.DataFrame(cosine_similarity(model), titles, titles)
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
df

stjorn1   stjorn2   stjorn3  \
stjorn1                                    1.000000  0.675328  0.782948   
stjorn2                                    0.675328  1.000000  0.653913   
stjorn3                                    0.782948  0.653913  1.000000   
stjorn4                                    0.649667  0.628012  0.617558   
nhb                                        0.492547  0.437319  0.558764   
nraNorrFragm75_kross_saga                  0.257317  0.252904  0.283761   
am132_egils_saga                           0.672262  0.603985  0.782889   
am162btheta_njals_saga                     0.571460  0.493016  0.650656   
nraNorrFragm64_barlaams_saga               0.499752  0.458419  0.557949   
nraNorrFragm81A_benedikts_regla            0.223068  0.198575  0.266781   
am1056IX_konungs_skuggsja_fragment         0.368039  0.325310  0.404185   
am78_kristinrettir                         0.420949  0.412752  0.492523   
am63_heimskringla3                         0.649515  0.567413  0.738019   
dg4-7_strengleikar                         0.728523  0.627521  0.817112   
am132_droplaugasona_saga                   0.581542  0.540146  0.672017   
am132_kormaks_saga                         0.476421  0.438389  0.549945   
nraNorrFragm72x76_dialogar                 0.493921  0.442038  0.562407   
nraNorrFragm53_haralds_saga_hardrada       0.391115  0.342973  0.476364   
am132_finnboga_saga                        0.575154  0.500493  0.649490   
nraNorrFragm70_agotu_saga                  0.233183  0.214355  0.273746   
nraNorrFragm62_karlamagnuss_saga           0.597920  0.544417  0.686656   
nraNorrFragm60A_stjorn                     0.356604  0.280096  0.348187   
am132_fostbraedra_saga                     0.627393  0.564595  0.700427   
lbsFragm82_olafs_saga_helga                0.573341  0.507356  0.691777   
nraNorrFragm58B_konungs_skuggsja           0.473809  0.431040  0.504335   
nraNorrFragm60C_stjorn                     0.484688  0.390875  0.590596   
holmPerg30_landslog                        0.361858  0.373689  0.437523   
am619_norwegian_homily_book                0.631445  0.574867  0.745470   
nhb_am619                                  0.631445  0.574867  0.745470   
nraNorrFragm57_jons_saga_helga             0.605003  0.529471  0.685471   
nraNorrFragm69_nikulass_saga               0.623527  0.518215  0.685765   
am56_landslog                              0.423865  0.411043  0.500364   
wolfAug9-10_egils_saga                     0.651869  0.591961  0.769610   
nraNorrFragm66_thomass_saga                0.601886  0.488404  0.642496   
holmPerg17_thomass_saga                    0.628796  0.463667  0.665807   
am383I_thorlaks_saga                       0.423834  0.341696  0.465623   
holmPerg4_thidreks_saga                    0.690059  0.641929  0.810510   
am132_njals_saga                           0.675828  0.622268  0.778592   
am36_heimskringla2                         0.669885  0.577155  0.767666   
am544_voluspa                              0.310470  0.285135  0.343586   
am162bkappa_njals_saga                     0.519042  0.484930  0.560267   
am305_landslog                             0.412932  0.399026  0.462660   
nraNorrFragm58C_konungs_skuggsja           0.560830  0.527444  0.619055   
am132_olkofra_thattr                       0.506382  0.482172  0.606494   
konungs_skuggsja_am243ba                   0.554423  0.503262  0.665866   
nraNorrFragm54_sverris_saga                0.668055  0.615986  0.743138   
nraNorrFragm55B_hakonar_saga               0.431877  0.404820  0.487555   
nraNorrFragm79_mariu_saga                  0.444717  0.391301  0.508651   
gks2365_voluspa                            0.328600  0.312088  0.366642   
am243balpha_konungs_skuggsja               0.554423  0.503262  0.665866   
nraNorrFragm51_fagrskinna                  0.385533  0.357703  0.414449   
am132_viga-glums_saga                      0.655217  0.596255  0.750953   
am279a_gragas                              0.341152  0.313256  0.3850

After substantial normalization, the two editions of NHB come to 0.69 similarity, which is still concerning, even if it is 15 percentage points better than before collapsing vowel length and þ/ð. There's probably Menota elements to disable yet, along the lines of `<sic>` and `<note>`. Really we expect a match in the 0.9-range.